src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,95 @@
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
-    public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+    public Object getInstance(Class<?> interfaceClass) {
+        Class<?> currentClass = findImplementation(interfaceClass);
+        Object instance = createNewInstance(currentClass);
+        initializeDependencies(instance);
+        return instance;
+    }
+
+    private Class<?> findImplementation(Class<?> interfaceOrComponentClass) {
+        if (interfaceOrComponentClass.isInterface()) {
+            return findComponentImplementation(interfaceOrComponentClass);
+        } else if (isComponent(interfaceOrComponentClass)) {
+            return interfaceOrComponentClass;
+        } else {
+            throw new RuntimeException("Class " + interfaceOrComponentClass.getName()
+                    + " is not annotated with @Component");
+        }
+    }
+
+    private Class<?> findComponentImplementation(Class<?> interfaceClass) {
+        Map<Class<?>, Class<?>> interfaceImplementations = new HashMap<>();
+        interfaceImplementations.put(FileReaderService.class, FileReaderServiceImpl.class);
+        interfaceImplementations.put(ProductParser.class, ProductParserImpl.class);
+        interfaceImplementations.put(ProductService.class, ProductServiceImpl.class);
+
+        Class<?> implementationClass = interfaceImplementations.get(interfaceClass);
+        if (implementationClass == null) {
+            throw new RuntimeException("No implementation found for interface "
+                    + interfaceClass.getName());
+        }
+
+        if (!isComponent(implementationClass)) {
+            throw new RuntimeException("Implementation class "
+                    + implementationClass.getName()
+                    + " is not annotated with @Component");
+        }
+
+        return implementationClass;
+    }
+
+    private boolean isComponent(Class<?> clazz) {
+        return clazz.isAnnotationPresent(Component.class);
+    }
+
+    private Object createNewInstance(Class<?> currentClass) {
+        if (instances.containsKey(currentClass)) {
+            return instances.get(currentClass);
+        }
+
+        try {
+            Constructor<?> constructor = currentClass.getConstructor();
+            Object instance = constructor.newInstance();
+            instances.put(currentClass, instance);
+            return instance;
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException("Can't create instance of " + currentClass.getName());
```

Comment:
It is better to replace many exceptions that have a common parent with a general parental exception. In this case, you can catch `ReflectiveOperationException` instead of listing all reflective exceptions.